### PR TITLE
Re-implement sys-clk-Editor using a state machine

### PR DESCRIPTION
**Pull request by cathery**
sun i'm so sorry

- Switch to a state machine similar to SPR manager
- Add multiple program states besides Enabled and Disabled (similar to SPR manager)
- Change "Delete config" to "Clear config", and actually clear the config

i haven't tested it completely, i don't remember how your gameSelected and configSelected variables are supposed to behave, please test to make sure everything works the same

### CODE REVIEW

**Comment by gabehxd** [comment]
with a state machine gameSelected and configSelected should no longer be necessary afaik as they were used to return to the correct selection on the last page

File: source/States/ConfigMenu.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+#include "StateMachine.h"
+
+class ConfigMenu : public State
+{
+public:
+    virtual void Enter() override {}
```
Comment (gabehxd) [inline source/States/ConfigMenu.h]: virtuals are redundant

**Review by gabehxd — CHANGES_REQUESTED** [review]

File: source/States/ClockMenus.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#pragma once
+#include "StateMachine.h"
+
+class MEMMenu : public State
+{
+public:
+    virtual void Enter() override {}
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    void Enter() override {}
```

File: source/States/ClockMenus.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    virtual void Update(StateMachine &, u64) override;
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    void Update(StateMachine &, u64) override;
```

File: source/States/ClockMenus.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    virtual void Exit() override {}
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    void Exit() override {}
```

File: source/States/ClockMenus.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    virtual std::string Name() override { return "mem"; }
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    std::string Name() override { return "mem"; }
```

File: source/States/ClockMenus.h [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    virtual ~MEMMenu() override {}
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    ~MEMMenu() override {}
```

File: source/States/ClockMenus.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+private:
+    int selection = 0;
+};
+
+class GPUMenu : public State
+{
+public:
+    virtual void Enter() override {}
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    void Enter() override {}
```

File: source/States/ClockMenus.h [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    virtual void Update(StateMachine &, u64) override;
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    void Update(StateMachine &, u64) override;
```

File: source/States/ClockMenus.h [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    virtual void Exit() override {}
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    void Exit() override {}
```

File: source/States/ClockMenus.h [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+    virtual std::string Name() override { return "gpu"; }
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    std::string Name() override { return "gpu"; }
```

File: source/States/ClockMenus.h [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+    virtual ~GPUMenu() override {}
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    ~GPUMenu() override {}
```

File: source/States/ClockMenus.h [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+private:
+    int selection = 0;
+};
+
+class CPUMenu : public State
+{
+public:
+    virtual void Enter() override {}
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    void Enter() override {}
```

File: source/States/ClockMenus.h [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    virtual void Update(StateMachine &, u64) override;
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    void Update(StateMachine &, u64) override;
```

File: source/States/ClockMenus.h [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    virtual void Exit() override {}
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    void Exit() override {}
```

File: source/States/ClockMenus.h [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+    virtual std::string Name() override { return "cpu"; }
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    std::string Name() override { return "cpu"; }
```

File: source/States/ClockMenus.h [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+    virtual ~CPUMenu() override {}
```
Comment (cathery) [inline source/States/ClockMenus.h]:
```suggestion
    ~CPUMenu() override {}
```

File: source/States/ConfigMenu.h (hunk 1 above)
Comment (cathery) [inline source/States/ConfigMenu.h]:
```suggestion
    void Enter() override {}
```

File: source/States/ConfigMenu.h [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+    virtual void Update(StateMachine &, u64) override;
```
Comment (cathery) [inline source/States/ConfigMenu.h]:
```suggestion
    void Update(StateMachine &, u64) override;
```

File: source/States/ConfigMenu.h [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+    virtual void Exit() override {}
```
Comment (cathery) [inline source/States/ConfigMenu.h]:
```suggestion
    void Exit() override {}
```

File: source/States/ConfigMenu.h [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+
+    virtual std::string Name() override { return "config"; }
```
Comment (cathery) [inline source/States/ConfigMenu.h]:
```suggestion
    std::string Name() override { return "config"; }
```

File: source/States/ConfigMenu.h [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+
+    virtual ~ConfigMenu() override {}
```
Comment (cathery) [inline source/States/ConfigMenu.h]:
```suggestion
    ~ConfigMenu() override {}
```

File: source/States/ErrorMenu.h [hunk 21]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+#include "StateMachine.h"
+
+class ErrorMenu : public State
+{
+    virtual void Enter() override {}
```
Comment (cathery) [inline source/States/ErrorMenu.h]:
```suggestion
    void Enter() override {}
```

File: source/States/ErrorMenu.h [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+    virtual void Update(StateMachine &, u64) override;
```
Comment (cathery) [inline source/States/ErrorMenu.h]:
```suggestion
    void Update(StateMachine &, u64) override;
```

File: source/States/ErrorMenu.h [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+    virtual void Exit() override {}
```
Comment (cathery) [inline source/States/ErrorMenu.h]:
```suggestion
    void Exit() override {}
```

File: source/States/ErrorMenu.h [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+
+    virtual std::string Name() override { return "error"; }
```
Comment (cathery) [inline source/States/ErrorMenu.h]:
```suggestion
    std::string Name() override { return "error"; }
```

File: source/States/ErrorMenu.h [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+
+    
+
+    virtual ~ErrorMenu() override {}
```
Comment (cathery) [inline source/States/ErrorMenu.h]:
```suggestion
    ~ErrorMenu() override {}
```

File: source/States/MainMenu.h [hunk 26]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+#include "StateMachine.h"
+
+class MainMenu : public State
+{
+public:
+    virtual void Enter() override;
```
Comment (cathery) [inline source/States/MainMenu.h]:
```suggestion
    void Enter() override;
```

File: source/States/MainMenu.h [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+    virtual void Update(StateMachine &, u64) override;
```
Comment (cathery) [inline source/States/MainMenu.h]:
```suggestion
    void Update(StateMachine &, u64) override;
```

File: source/States/MainMenu.h [hunk 28]
```diff
@@ -0,0 +1,19 @@
...
+    virtual void Exit() override {}
```
Comment (cathery) [inline source/States/MainMenu.h]:
```suggestion
    void Exit() override {}
```

File: source/States/MainMenu.h [hunk 29]
```diff
@@ -0,0 +1,19 @@
...
+
+    virtual std::string Name() override { return "main"; }
```
Comment (cathery) [inline source/States/MainMenu.h]:
```suggestion
    std::string Name() override { return "main"; }
```

File: source/States/MainMenu.h [hunk 30]
```diff
@@ -0,0 +1,19 @@
...
+
+    virtual ~MainMenu() override {}
```
Comment (cathery) [inline source/States/MainMenu.h]:
```suggestion
    ~MainMenu() override {}
```

File: source/States/TitleMenu.h [hunk 31]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+#include "StateMachine.h"
+
+class TitleMenu : public State
+{
+public:
+    virtual void Enter() override {}
```
Comment (cathery) [inline source/States/TitleMenu.h]:
```suggestion
    void Enter() override {}
```

File: source/States/TitleMenu.h [hunk 32]
```diff
@@ -0,0 +1,17 @@
...
+    virtual void Update(StateMachine &, u64) override;
```
Comment (cathery) [inline source/States/TitleMenu.h]:
```suggestion
    void Update(StateMachine &, u64) override;
```

File: source/States/TitleMenu.h [hunk 33]
```diff
@@ -0,0 +1,17 @@
...
+    virtual void Exit() override {}
```
Comment (cathery) [inline source/States/TitleMenu.h]:
```suggestion
    void Exit() override {}
```

File: source/States/TitleMenu.h [hunk 34]
```diff
@@ -0,0 +1,17 @@
...
+
+    virtual std::string Name() override { return "title"; }
```
Comment (cathery) [inline source/States/TitleMenu.h]:
```suggestion
    std::string Name() override { return "title"; }
```

File: source/States/TitleMenu.h [hunk 35]
```diff
@@ -0,0 +1,17 @@
...
+
+    virtual ~TitleMenu() override {}
```
Comment (cathery) [inline source/States/TitleMenu.h]:
```suggestion
    ~TitleMenu() override {}
```

**Review by gabehxd — APPROVED** [review]

File: source/States/ClockMenus.cpp [hunk 36]
```diff
@@ -0,0 +1,103 @@
...
+    if (kDown & KEY_B)
+    {
+        stateMachine.PushState("config");
+        return;
+    }
+}
+
+void GPUMenu::Update(StateMachine &stateMachine, u64 kDown)
+{
+    if (kDown & KEY_UP)
```
Comment (gabehxd) [inline source/States/ClockMenus.cpp:35]: nitpick but single line these?

File: source/States/ClockMenus.h [hunk 37]
```diff
@@ -0,0 +1,47 @@
+#pragma once
+#include "StateMachine.h"
+
+class MEMMenu : public State
+{
+public:
+    void Enter() override {}
```
Comment (gabehxd) [inline source/States/ClockMenus.h]: empty implementations are not necessary for overrides as they are virtual

File: source/States/ErrorMenu.h [hunk 38]
```diff
@@ -0,0 +1,18 @@
...
+#include "StateMachine.h"
+
+class ErrorMenu : public State
+{
+    void Enter() override {}
+    void Update(StateMachine &, u64) override;
+    void Exit() override {}
+
+    std::string Name() override { return "error"; }
+
```
Comment (gabehxd) [inline source/States/ErrorMenu.h]: lots of extra spaces

**Review by gabehxd — CHANGES_REQUESTED** [review]

**Comment by gabehxd** [comment]
just nitpicking really :p

**Review by gabehxd — APPROVED** [review]

**Comment by gabehxd** [comment]
all seems good in testing 👌

**Pull request merged**